File: app/src/main/java/com/khmelenko/lab/varis/widget/BuildView.java
package com.khmelenko.lab.varis.widget;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.khmelenko.lab.varis.R;
import com.khmelenko.lab.varis.converter.BuildStateHelper;
import com.khmelenko.lab.varis.converter.TimeConverter;
import com.khmelenko.lab.varis.network.response.Commit;
import com.khmelenko.lab.varis.network.response.IBuildState;
import com.khmelenko.lab.varis.network.response.RequestData;
import com.khmelenko.lab.varis.util.DateTimeUtils;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * View with build info
 *
 * @author Dmytro Khmelenko
 */
public class BuildView extends LinearLayout {

    @BindView(R.id.build_indicator)
    View mIndicator;

    @BindView(R.id.build_number)
    TextView mTitle;

    @BindView(R.id.build_pull_request_title)
    TextView mPullRequest;

    @BindView(R.id.build_branch)
    TextView mBranch;

    @BindView(R.id.build_commit_message)
    TextView mCommitMessage;

    @BindView(R.id.build_commit_person)
    TextView mCommitPerson;

    @BindView(R.id.build_duration)
    TextView mDuration;

    @BindView(R.id.build_finished)
    TextView mFinished;

    public BuildView(Context context) {
        super(context);
        initializeViews(context);
    }

    public BuildView(Context context, AttributeSet attrs) {
        super(context, attrs);
        initializeViews(context);
    }

    public BuildView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        initializeViews(context);
    }

    /**
     * Inflates the views in the layout.
     *
     * @param context the current context for the view.
     */
    private void initializeViews(Context context) {
        LayoutInflater inflater = (LayoutInflater) context
                .getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        View view = inflater.inflate(R.layout.view_build, this);
        ButterKnife.bind(this, view);
    }

    /**
     * Sets build duration
     *
     * @param durationInMillis Duration
     */
    public void setDuration(long durationInMillis) {
        if (durationInMillis != 0) {
            String duration = TimeConverter.durationToString(durationInMillis);
            duration = getContext().getString(R.string.build_duration, duration);
            mDuration.setText(duration);
            mDuration.setVisibility(VISIBLE);
        } else {
            mDuration.setVisibility(GONE);
        }
    }

    /**
     * Sets the message when the build was finished
     *
     * @param finishedAt String with the message when the build was finished
     */
    public void setFinishedAt(String finishedAt) {
        if (!TextUtils.isEmpty(finishedAt)) {
            String formattedDate = DateTimeUtils.parseAndFormatDateTime(finishedAt);
            formattedDate = getContext().getString(R.string.build_finished_at, formattedDate);
            mFinished.setText(formattedDate);
            mFinished.setVisibility(VISIBLE);
        } else {
            mFinished.setVisibility(GONE);
        }
    }

    /**
     * Sets build state
     *
     * @param buildState Build state
     */
    public void setState(IBuildState buildState) {
        if (buildState != null) {
            setTitle(getContext()
                    .getString(R.string.build_build_number, buildState.getNumber(), buildState.getState()));
            setStateIndicator(buildState.getState());
            setFinishedAt(buildState.getFinishedAt());
            setDuration(buildState.getDuration());
        }
    }

    /**
     * Sets commit data
     *
     * @param commit Last commit
     */
    public void setCommit(Commit commit) {
        if (commit != null) {
            mBranch.setVisibility(VISIBLE);
            mBranch.setText(commit.getBranch());
            mCommitMessage.setVisibility(VISIBLE);
            mCommitMessage.setText(commit.getMessage());
            mCommitPerson.setVisibility(VISIBLE);
            mCommitPerson.setText(commit.getCommitterName());
        }
    }

    /**
     * Sets title
     *
     * @param title Title
     */
    public void setTitle(String title) {
        mTitle.setText(title);
    }

    /**
     * Sets state indicator
     *
     * @param state Build state
     */
    public void setStateIndicator(String state) {
        if (!TextUtils.isEmpty(state)) {
            int buildColor = BuildStateHelper.getBuildColor(state);
            mIndicator.setBackgroundColor(buildColor);
            mTitle.setTextColor(buildColor);

            Drawable drawable = BuildStateHelper.getBuildImage(state);
            if (drawable != null) {
                mTitle.setCompoundDrawablesWithIntrinsicBounds(drawable, null, null, null);
            }
        }
    }

    /**
     * Sets the title for pull request
     *
     * @param title Title for Pull Request
     */
    public void setPullRequestTitle(RequestData title) {
        mPullRequest.setVisibility(VISIBLE);
        mPullRequest.setText(title.getPullRequestTitle());
    }
}

File: app-v3/src/main/java/com/khmelenko/lab/varis/widget/BuildView.kt
package com.khmelenko.lab.varis.widget

import android.content.Context
import android.text.TextUtils
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.LinearLayout
import com.khmelenko.lab.varis.R
import com.khmelenko.lab.varis.converter.BuildStateHelper
import com.khmelenko.lab.varis.converter.TimeConverter
import com.khmelenko.lab.varis.network.response.Commit
import com.khmelenko.lab.varis.network.response.IBuildState
import com.khmelenko.lab.varis.network.response.RequestData
import com.khmelenko.lab.varis.util.DateTimeUtils
import kotlinx.android.synthetic.main.view_build.view.build_branch
import kotlinx.android.synthetic.main.view_build.view.build_commit_message
import kotlinx.android.synthetic.main.view_build.view.build_commit_person
import kotlinx.android.synthetic.main.view_build.view.build_duration
import kotlinx.android.synthetic.main.view_build.view.build_finished
import kotlinx.android.synthetic.main.view_build.view.build_indicator
import kotlinx.android.synthetic.main.view_build.view.build_number
import kotlinx.android.synthetic.main.view_build.view.build_pull_request_title

/**
 * View with build info
 *
 * @author Dmytro Khmelenko
 */
class BuildView : LinearLayout {

    constructor(context: Context) : super(context) {
        initializeViews(context)
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        initializeViews(context)
    }

    constructor(context: Context, attrs: AttributeSet, defStyle: Int) : super(context, attrs, defStyle) {
        initializeViews(context)
    }

    /**
     * Inflates the views in the layout.
     *
     * @param context the current context for the view.
     */
    private fun initializeViews(context: Context) {
        val inflater = context
                .getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        inflater.inflate(R.layout.view_build, this)
    }

    /**
     * Sets build duration
     *
     * @param durationInMillis Duration
     */
    fun setDuration(durationInMillis: Long) {
        if (durationInMillis != 0L) {
            var duration = TimeConverter.durationToString(durationInMillis)
            duration = context.getString(R.string.build_duration, duration)
            build_duration.text = duration
            build_duration.visibility = View.VISIBLE
        } else {
            build_duration.visibility = View.GONE
        }
    }

    /**
     * Sets the message when the build was finished
     *
     * @param finishedAt String with the message when the build was finished
     */
    fun setFinishedAt(finishedAt: String) {
        if (!TextUtils.isEmpty(finishedAt)) {
            var formattedDate = DateTimeUtils.parseAndFormatDateTime(finishedAt)
            formattedDate = context.getString(R.string.build_finished_at, formattedDate)
            build_finished.text = formattedDate
            build_finished.visibility = View.VISIBLE
        } else {
            build_finished.visibility = View.GONE
        }
    }

    /**
     * Sets build state
     *
     * @param buildState Build state
     */
    fun setState(buildState: IBuildState?) {
        if (buildState != null) {
            setTitle(context
                    .getString(R.string.build_build_number, buildState.number, buildState.state))
            setStateIndicator(buildState.state)
            setFinishedAt(buildState.finishedAt)
            setDuration(buildState.duration)
        }
    }

    /**
     * Sets commit data
     *
     * @param commit Last commit
     */
    fun setCommit(commit: Commit?) {
        if (commit != null) {
            build_branch.visibility = View.VISIBLE
            build_branch.text = commit.branch
            build_commit_message.visibility = View.VISIBLE
            build_commit_message.text = commit.message
            build_commit_person.visibility = View.VISIBLE
            build_commit_person.text = commit.committerName
        }
    }

    /**
     * Sets title
     *
     * @param title Title
     */
    fun setTitle(title: String) {
        build_number.text = title
    }

    /**
     * Sets state indicator
     *
     * @param state Build state
     */
    fun setStateIndicator(state: String) {
        if (!TextUtils.isEmpty(state)) {
            val buildColor = BuildStateHelper.getBuildColor(state)
            build_indicator.setBackgroundColor(buildColor)
            build_number.setTextColor(buildColor)

            val drawable = BuildStateHelper.getBuildImage(state)
            if (drawable != null) {
                build_number.setCompoundDrawablesWithIntrinsicBounds(drawable, null, null, null)
            }
        }
    }

    /**
     * Sets the title for pull request
     *
     * @param title Title for Pull Request
     */
    fun setPullRequestTitle(title: RequestData) {
        build_pull_request_title.visibility = View.VISIBLE
        build_pull_request_title.text = title.pullRequestTitle
    }
}
